storm-core/src/jvm/org/apache/storm/scheduler/blacklist/BlacklistScheduler.java

```diff
@@ -0,0 +1,252 @@
+                return blacklistHost.size();
+            }
+        });
+    }
+
+    @Override
+    public void schedule(Topologies topologies, Cluster cluster) {
+        LOG.debug("running Black List scheduler");
+        Map<String, SupervisorDetails> supervisors = cluster.getSupervisors();
+        for (Map.Entry<String, SupervisorDetails> entry : supervisors.entrySet()) {
+            SupervisorDetails supervisorDetails = entry.getValue();
+            String hosts = supervisorDetails.getHost();
+            Set<Integer> ports = supervisorDetails.getAllPorts();
+            LOG.debug("supervisor: {}, ports: {}", hosts, ports);
+        }
+        LOG.debug("AssignableSlots: {}", cluster.getAssignableSlots());
+        LOG.debug("AvailableSlots: {}", cluster.getAvailableSlots());
+        LOG.debug("UsedSlots: {}", cluster.getUsedSlots());
+
+        blacklistStrategy.resumeFromBlacklist();
+        badSupervisors(supervisors);
+        cluster.setBlacklistedHosts(getBlacklistHosts(cluster, topologies));
+        removeLongTimeDisappearFromCache();
+
+        underlyingScheduler.schedule(topologies, cluster);
+    }
+
+    private void badSupervisors(Map<String, SupervisorDetails> supervisors) {
+        Set<String> cachedSupervisorsKeySet = cachedSupervisors.keySet();
+        Set<String> supervisorsKeySet = supervisors.keySet();
+
+        Set<String> badSupervisorKeys = Sets.difference(cachedSupervisorsKeySet, supervisorsKeySet);//cached supervisor doesn't show up
+        HashMap<String, Set<Integer>> badSupervisors = new HashMap<String, Set<Integer>>();
+        for (String key : badSupervisorKeys) {
+            badSupervisors.put(key, cachedSupervisors.get(key));
+        }
+
+        for (Map.Entry<String, SupervisorDetails> entry : supervisors.entrySet()) {
+            String key = entry.getKey();
+            SupervisorDetails supervisorDetails = entry.getValue();
+            if (cachedSupervisors.containsKey(key)) {
+                Set<Integer> badSlots = badSlots(supervisors, key);
+                if (badSlots.size() > 0) {//supervisor contains bad slots
+                    badSupervisors.put(key, badSlots);
+                }
+            } else {
+                cachedSupervisors.put(key, supervisorDetails.getAllPorts());//new supervisor to cache
+            }
+        }
+
+        badSupervisorsTolerance.add(badSupervisors);
+    }
+
+    private Set<Integer> badSlots(Map<String, SupervisorDetails> supervisors, String supervisorKey) {
+        SupervisorDetails supervisor = supervisors.get(supervisorKey);
+        Set<Integer> cachedSupervisorPorts = cachedSupervisors.get(supervisorKey);
+        Set<Integer> supervisorPorts = supervisor.getAllPorts();
+
+        Set<Integer> newPorts = Sets.difference(supervisorPorts, cachedSupervisorPorts);
+        if (newPorts.size() > 0) {
+            cachedSupervisors.put(supervisorKey, Sets.union(newPorts, supervisor.getAllPorts()));
```

Comment:
shouldn't the value here be just supervisor.getAllPorts()? The union doesn't hurt, just I think "newPorts" is a subset of supervisor.getAllPorts.